File: homework4/steven_law/q1_BuildATrie.cpp
/*Data Structure: Trie
Algorithm: prefix tree — insert/search walk the path, remove unwinds + prunes dead nodes
Time Complexity: insert O(L), isValidWord O(L), remove O(L)   (L = word length)
Space Complexity: O(total chars inserted)
Time Taken: 35 mins 12 seconds
*/

#include<iostream>
#include<vector>
#include <string>

using namespace std;

/*
1. node = 26 children + validWord flag. root is an empty node.
2. insert: walk letter by letter, make the child if it doesnt exist, flag last node validWord.
3. isValidWord: walk the path, if any child missing return false, else return last->validWord.
4. remove: recurse to the end, unflag validWord, then on the way back delete any node that
   has no children AND isnt a word end. never delete root.
*/

struct TrieNode {
    vector<TrieNode*> children;
    bool validWord;
    TrieNode() : children(26, nullptr), validWord(false) {}
};

class Trie {
    TrieNode* root;

    //returns true if the caller should delete this node (no kids + not a word end)
    bool removeHelper(TrieNode* node, const string& word, int i){
        if(node == nullptr) return false;

        if(i == (int)word.size()){
            if(!node->validWord) return false; //word was never in the trie
            node->validWord = false;
        } else {
            int idx = word[i] - 'a';
            if(removeHelper(node->children[idx], word, i+1)){
                delete node->children[idx];
                node->children[idx] = nullptr;
            }
        }

        if(node->validWord) return false;
        for(auto& c : node->children) if(c != nullptr) return false;
        return node != root; //dont let root get deleted
    }

public:
    Trie() : root(new TrieNode()) {}

    void insert(string word){
        TrieNode* curr = root;
        for(char ch : word){
            int idx = ch - 'a';
            if(curr->children[idx] == nullptr) curr->children[idx] = new TrieNode();
            curr = curr->children[idx];
        }
        curr->validWord = true;
    }

    bool isValidWord(string word){
        TrieNode* curr = root;
        for(char ch : word){
            int idx = ch - 'a';
            if(curr->children[idx] == nullptr) return false;
            curr = curr->children[idx];
        }
        return curr->validWord;
    }

    void remove(string word){
        removeHelper(root, word, 0);
    }
};

int main(){
    Trie t;
    t.insert("cat");
    t.insert("car");
    t.insert("care");

    cout << boolalpha;
    cout << "isValidWord(cat)  expect true  -> " << t.isValidWord("cat")  << endl;
    cout << "isValidWord(ca)   expect false -> " << t.isValidWord("ca")   << endl;
    cout << "isValidWord(care) expect true  -> " << t.isValidWord("care") << endl;

    t.remove("car");
    cout << "after remove(car):" << endl;
    cout << "isValidWord(car)  expect false -> " << t.isValidWord("car")  << endl;
    cout << "isValidWord(care) expect true  -> " << t.isValidWord("care") << endl; //'car' path must survive
    cout << "isValidWord(cat)  expect true  -> " << t.isValidWord("cat")  << endl;
}

/* psudocode/thoughts/logic
goal: trie with insert, isValidWord (search), remove (and clean up unused nodes)
restraints: removing a word cant break other words that share its path (remove "car" must
keep "care" alive)

strategy: each node is 26 children + a validWord bool. insert/search are simple walks. remove
is the tricky one — recurse down, clear the validWord at the end, then on the unwind delete a
child only if its now empty (no children) and isnt itself a word end. the node != root guard
stops us from ever nuking the root.
*/

File: homework4/yonathan_abera/q1_BuildATrie.py
# Data Structure: Trie
# Time Complexity: O(L) per insert/search/remove
# Space Complexity: O(L) per inserted word

class TrieNode:
    def __init__(self):
        self.children = [None] * 26
        self.valid_word = False

    def has_children(self):
        return any(self.children)


class Trie:
    def __init__(self):
        self.root = TrieNode()

    def insert(self, word):
        node = self.root
        for ch in word:
            i = ord(ch) - ord('a')
            if node.children[i] is None:
                node.children[i] = TrieNode()
            node = node.children[i]
        node.valid_word = True

    def isValidWord(self, word):
        node = self.root
        for ch in word:
            i = ord(ch) - ord('a')
            if node.children[i] is None:
                return False
            node = node.children[i]
        return node.valid_word

    def remove(self, word):
        self._remove(self.root, word, 0)

    def _remove(self, node, word, depth):
        if node is None:
            return False
        if depth == len(word):
            node.valid_word = False
            return not node.has_children()
        i = ord(word[depth]) - ord('a')
        if self._remove(node.children[i], word, depth + 1):
            node.children[i] = None
        return not node.valid_word and not node.has_children()

t = Trie()
for w in ["cat", "car", "card", "dog"]:
    t.insert(w)
print(t.isValidWord("cat"))
print(t.isValidWord("ca"))
print(t.isValidWord("cab"))
t.remove("car")
print(t.isValidWord("car"))
print(t.isValidWord("card"))
t.remove("zebra")
print(t.isValidWord("dog"))

# Time spent: 30
